slashing/program/src/processor.rs

```diff
@@ -0,0 +1,138 @@
+//! Program state processor
+
+use {
+    crate::{error::SlashingError, instruction::SlashingInstruction, state::ProofData},
+    solana_program::{
+        account_info::{next_account_info, AccountInfo},
+        entrypoint::ProgramResult,
+        msg,
+        program::invoke,
+        program_error::ProgramError,
+        program_pack::IsInitialized,
+        pubkey::Pubkey,
+        rent::Rent,
+        system_instruction, system_program,
+    },
+    spl_pod::bytemuck::{pod_from_bytes, pod_from_bytes_mut},
+};
+
+fn check_authority(authority_info: &AccountInfo, expected_authority: &Pubkey) -> ProgramResult {
+    if expected_authority != authority_info.key {
+        msg!("Incorrect proof authority provided");
+        return Err(SlashingError::IncorrectAuthority.into());
+    }
+    if !authority_info.is_signer {
+        msg!("Proof authority signature missing");
+        return Err(ProgramError::MissingRequiredSignature);
+    }
+    Ok(())
+}
+
+/// Instruction processor
+pub fn process_instruction(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    input: &[u8],
+) -> ProgramResult {
+    let instruction = SlashingInstruction::unpack(input)?;
+    let account_info_iter = &mut accounts.iter();
+
+    match instruction {
+        SlashingInstruction::InitializeProofAccount { proof_type } => {
+            msg!(
+                "SlashingInstruction::InitializeProofAccount {:?}",
+                proof_type
+            );
+
+            let proof_data_info = next_account_info(account_info_iter)?;
+            let payer_info = next_account_info(account_info_iter)?;
+            let authority_info = next_account_info(account_info_iter)?;
+            let system_program_info = next_account_info(account_info_iter)?;
+            let account_length = proof_type.proof_account_length()?;
+
+            if *system_program_info.key != system_program::ID {
+                msg!("Missing system program account");
+                return Err(ProgramError::InvalidAccountData);
+            }
+
+            msg!("Creating proof account with size {}", account_length);
```

Comment:
I provided the functionality for the user to create and initialize the account here, but if you prefer we could have them create the account and write it using themselves using the record program, only interacting with slashing for the verification.

Not sure what the preferred style is for these things.

Reply:
That's a good question. If these proofs might be really big, we could go with the same model as zk proofs, where the proofs can be read from instruction data or account data (doesn't matter which program). 

This way, the slashing program gets way simpler -- it just needs to take in data, check it, and then create a valid slash "context state account" from that, containing the minimum required data, ie:

* validator that committed the slashable offense
* pubkey that submitted the proof
* kind of slashable offense
* epoch in which it happened
 
These would look similar to the "context state accounts" in the zk elgamal proof program. Here's how the program looks: https://github.com/anza-xyz/agave/blob/master/programs/zk-elgamal-proof/src/lib.rs

And the `ZkProofData` trait: https://github.com/anza-xyz/agave/blob/master/zk-sdk/src/zk_elgamal_proof_program/proof_data/mod.rs#L65

And one example separating the proof data from context: https://github.com/anza-xyz/agave/blob/c88e6df566c5c17d71e9574785755683a8fb033a/zk-sdk/src/zk_elgamal_proof_program/proof_data/pubkey_validity.rs#L35

The only issue with using account data owned by any program is that the data could get changed afterwards, which can make auditability more difficult. The slashing program could get around that by using the same trick as account-compression, where the program CPIs into a no-op program with all of the proof data, which forces the runtime to log the instruction data. This way it's stored longterm.

Or, keep it simple, and make the "context state account" contain all the data.

Does this make sense? What do you think?

Reply:
That's fair, I am not opposed to having the user create and write the proof data themselves.
To answer your questions:
* The proof sizes will be fairly large. I expect the duplicate block proof to be the smallest with ~2400 bytes
* For now I expect the proofs to be of fixed size known up front for each proof type, although this might change if we decide to add more slashing conditions in the future
* There also should be no reason to CPI into the slashing program, or create the proof on chain. All slashable events occur (at least partially) offchain by nature.
* I haven't decided yet on how the results will be collected but the context state account makes sense. At most we'd need 1 account per staked validator. 
We could also just maintain a more compact structure instead of creating a separate account: `HashMap<Epoch, (Pubkey, Vec<ProofType, usize>>>` essentially a list of # of infractions by type per pubkey.

Reply:
Gotcha, that's all good context. A few followups:

> There also should be no reason to CPI into the slashing program, or create the proof on chain. All slashable events occur (at least partially) offchain by nature.

You never know what people will do! Just so I'm sure I'm not totally misunderstanding, when you say "no reason to create the proof on chain", you mean specifically "from another program", correct? Because the concept is that someone will recreate the whole proof on-chain, right?

>  I haven't decided yet on how the results will be collected but the context state account makes sense. At most we'd need 1 account per staked validator.

Ah right, yeah that makes sense. I was thinking (naively) that we wouldn't even need accounts per validator, since you can just submit proofs as individual accounts, but we probably want to avoid duplicate submissions. For example, if a validator creates a duplicate block, and they created more than one pair of offending shreds, then there will be two distinct valid slash proofs, but for the same event.

Ok, so then we'll need two account types: proofs (can be created on any address) and validator accounts (some PDA that can be instantiated by anyone). Let's avoid the hashmap -- that can become an easy attack vector.

Reply:
> you mean specifically "from another program", correct

Yeah exactly, the proof will still be uploaded on-chain.

> For example, if a validator creates a duplicate block, and they created more than one pair of offending shreds, then there will be two distinct valid slash proofs, but for the same event.

This is a good point, avoiding duplicates definitely makes it worth tracking per validator. I was also thinking that since the runtime code will eventually have to traverse these accounts in order to debit delegations it's a good idea to keep it bounded.

>Let's avoid the hashmap -- that can become an easy attack vector.

Sounds good.